File: src/lib/path.ts
/**
 * PathUtils — TS 镜像
 * 语义与 `src-tauri/src/algorithm/path.rs:1:1`
 *
 * 桌面端**不**用 SAF，路径就是普通字符串（绝对路径）。
 * 这模块只做字符串切片/拼接/面包屑构建，不做 OS 调用。
 */

/** 面包屑中的一项：显示名 + 累积的规范化路径 */
export interface Crumb {
  label: string;
  path: string;
}

/** 把路径切成 segments（按 `/` 或 `\` 分隔），去空段 */
function segmentsImpl(path: string): string[] {
  return path
    .replace(/\\/g, '/')
    .split('/')
    .filter((s) => s.length > 0);
}

/** 规范化路径：合并多余分隔符 */
function normalizeImpl(path: string): string {
  return segmentsImpl(path).join('/');
}

/** 拼接路径（base + segment） */
function joinImpl(base: string, segment: string): string {
  const normBase = normalizeImpl(base);
  const normSeg = normalizeImpl(segment);
  if (normBase.length === 0) return normSeg;
  if (normSeg.length === 0) return normBase;
  return `${normBase}/${normSeg}`;
}

/** 取父目录（移除最后一段；无父则返回空串） */
function parentImpl(path: string): string {
  const segs = segmentsImpl(path);
  if (segs.length <= 1) return '';
  return segs.slice(0, -1).join('/');
}

/** 面包屑（每段累计路径，第一项是根标签 + 空路径） */
function crumbsImpl(rootLabel: string, path: string): Crumb[] {
  const segs = segmentsImpl(path);
  const result: Crumb[] = [{ label: rootLabel, path: '' }];
  let acc = '';
  for (const seg of segs) {
    acc = acc.length === 0 ? seg : `${acc}/${seg}`;
    result.push({ label: seg, path: acc });
  }
  return result;
}

/** 命名空间导出（方便 `PathUtils.segments(...)` 调用），与 Rust `PathUtils` 镜像 */
export const PathUtils = {
  segments: segmentsImpl,
  normalize: normalizeImpl,
  join: joinImpl,
  parent: parentImpl,
  crumbs: crumbsImpl,
} as const;

File: src-tauri-algorithm-tests/src/path.rs
//! 路径工具（与 src-tauri/src/algorithm/path.rs 镜像）

pub fn segments(path: &str) -> Vec<String> {
    path.replace('\\', "/")
        .split('/')
        .filter(|s| !s.is_empty())
        .map(String::from)
        .collect()
}

pub fn normalize(path: &str) -> String {
    segments(path).join("/")
}

pub fn join(base: &str, segment: &str) -> String {
    let norm_base = normalize(base);
    let norm_seg = normalize(segment);
    if norm_base.is_empty() {
        norm_seg
    } else {
        format!("{}/{}", norm_base, norm_seg)
    }
}

pub fn parent(path: &str) -> String {
    let segs = segments(path);
    if segs.len() <= 1 {
        String::new()
    } else {
        segs[..segs.len() - 1].join("/")
    }
}

pub fn crumbs(root_label: &str, path: &str) -> Vec<(String, String)> {
    let segs = segments(path);
    let mut result = vec![(root_label.to_string(), String::new())];
    let mut acc = String::new();
    for seg in segs {
        acc = if acc.is_empty() {
            seg.clone()
        } else {
            format!("{}/{}", acc, seg)
        };
        result.push((seg, acc.clone()));
    }
    result
}

pub struct PathUtils;

impl PathUtils {
    pub fn segments(path: &str) -> Vec<String> { segments(path) }
    pub fn normalize(path: &str) -> String { normalize(path) }
    pub fn join(base: &str, segment: &str) -> String { join(base, segment) }
    pub fn parent(path: &str) -> String { parent(path) }
    pub fn crumbs(root_label: &str, path: &str) -> Vec<(String, String)> { crumbs(root_label, path) }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_segments() {
        assert_eq!(segments("a/b/c"), vec!["a", "b", "c"]);
        assert_eq!(segments("a\\b\\c"), vec!["a", "b", "c"]);
        assert_eq!(segments("/a/b/"), vec!["a", "b"]);
        assert_eq!(segments(""), Vec::<String>::new());
    }

    #[test]
    fn test_normalize() {
        assert_eq!(normalize("a//b\\\\c"), "a/b/c");
    }

    #[test]
    fn test_join() {
        assert_eq!(join("a/b", "c/d"), "a/b/c/d");
        assert_eq!(join("a/b", "/c/d"), "a/b/c/d");
        assert_eq!(join("", "a"), "a");
    }

    #[test]
    fn test_parent() {
        assert_eq!(parent("a/b/c"), "a/b");
        assert_eq!(parent("a"), "");
        assert_eq!(parent(""), "");
    }

    #[test]
    fn test_crumbs() {
        let c = crumbs("Root", "docs/comics");
        assert_eq!(c.len(), 3);
        assert_eq!(c[0], ("Root".to_string(), "".to_string()));
        assert_eq!(c[1], ("docs".to_string(), "docs".to_string()));
        assert_eq!(c[2], ("comics".to_string(), "docs/comics".to_string()));
    }
}